test/distributed/checkpoint/fsdp/test_fsdp_dsd.py

```diff
@@ -182,6 +187,392 @@ def _test_save_with_fsdp1_and_load_with_fsdp2(self, state_dict_type: StateDictTy
         self.assertEqual(fsdp2_full_msd, fsdp1_full_msd)
         self.assertEqual(fsdp1_full_osd, fsdp2_full_osd)
 
+    @skip_if_lt_x_gpu(4)
+    @with_temp_dir
+    def test_save_with_fsdp1_and_load_with_fsdp2_tp(self):
+        """
+        Test that we can save a model with FSDP1 and load it with FSDP2 + TP on 2d mesh.
+        """
+
+        def _get_base_model(mlp_dim: int = 2):
+            base_model = nn.Sequential(MLP(mlp_dim), MLP(mlp_dim), MLP(mlp_dim))
+            return base_model
+
+        state_dict_type_list = [
+            StateDictType.FULL_STATE_DICT,
+            StateDictType.SHARDED_STATE_DICT,
+        ]
+
+        for state_dict_type in state_dict_type_list:
+            # Save state dict with original model
+            base_model = _get_base_model().cuda()
+            base_optim = torch.optim.Adam(base_model.parameters(), lr=0.1)
+
+            # Save state dict with model wrapped with FSDP1
+            fsdp1_model = FSDP(
+                copy.deepcopy(base_model),
+                use_orig_params=True,
+                auto_wrap_policy=always_wrap_policy,
+            )
+
+            fsdp1_optim = torch.optim.Adam(fsdp1_model.parameters(), lr=0.1)
+
+            # one-step training to modify state dict
+            inp = torch.randn((2,), device=self.rank)
+            base_model(inp).sum().backward()
+            base_optim.step()
+            fsdp1_model(inp).sum().backward()
+            fsdp1_optim.step()
+
+            # obtain the unsharded state dict
+            base_msd = get_model_state_dict(
+                base_model,
+                options=StateDictOptions(full_state_dict=True, cpu_offload=True),
+            )
+            base_osd = get_optimizer_state_dict(
+                base_model,
+                base_optim,
+                options=StateDictOptions(full_state_dict=True, cpu_offload=True),
+            )
+
+            with FSDP.state_dict_type(fsdp1_model, state_dict_type):
+                fsdp1_state_dict = {
+                    "model": fsdp1_model.state_dict(),
+                    "optim": FSDP.sharded_optim_state_dict(fsdp1_model, fsdp1_optim),
+                }
```

Comment:
I tried the following and got `NotImplementedError: Only single local shard is supported.` error:
```
fsdp1_msd = get_model_state_dict(
    fsdp1_model,
    options=StateDictOptions(full_state_dict=save_full_state_dict), 
)
```

trace:
```
File "/data/users/xilunwu/oss/pytorch/test/distributed/checkpoint/fsdp/test_fsdp_dsd.py", line 238, in test_save_with_fsdp1_and_load_with_fsdp2_tp
    fsdp1_msd = get_model_state_dict(
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/checkpoint/state_dict.py", line 991, in get_model_state_dict
    model_state_dict = _get_model_state_dict(model, info)
  File "/data/users/xilunwu/oss/pytorch/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/checkpoint/state_dict.py", line 523, in _get_model_state_dict
    if torch.is_tensor(p) and p.is_meta:
  File "/home/xilunwu/local/miniconda3/envs/torchtitan/lib/python3.10/site-packages/typing_extensions.py", line 2636, in wrapper
    return arg(*args, **kwargs)
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/_shard/sharded_tensor/api.py", line 1168, in __torch_function__
    return dispatch(st_instance, func)
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/_shard/sharded_tensor/api.py", line 1149, in dispatch
    return _SHARDED_OPS[func](types, args, kwargs, st._process_group)
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/_shard/op_registry_utils.py", line 38, in wrapper
    return wrapped_func(types, args, kwargs, process_group)
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/_shard/sharded_tensor/_ops/tensor_ops.py", line 59, in st_is_meta
    return args[0].local_tensor().is_meta
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/_shard/sharded_tensor/api.py", line 1131, in local_tensor
    raise NotImplementedError("Only single local shard is supported.")
NotImplementedError: Only single local shard is supported.
```

Reply:
Oh. There is a corner case that ShardedTensor doesn't handle. TLDR is to pass a device_mesh to FSDP1 call so that the state dict would be in DTensor instead.

More context: https://github.com/pytorch/pytorch/issues/132366#issuecomment-2264642034

Reply:
Thanks @wz337 for the suggestion on using device_mesh argument, and it works!

Reply:
What I don't understand is that why @XilunWu didn't encounter such an issue when using FSDP1 native state_dict API. I'll verify if something is wrong with get_state_dict.